### PR TITLE
Class config

### PR DESCRIPTION
**Pull request by rhsimplex**
@josh for discussion - minimal delta for the configuration change w.r.t. #10

### CODE REVIEW

**Comment by jwayne** [comment]
This looks great to me. I just made some small modifications (~15min) to get this branch running on the following examples. I've manually tested that all visualizations work for these examples.

```
FLASK_APP=picasso PICASSO_SETTINGS=examples/keras/config.py flask run
FLASK_APP=picasso PICASSO_SETTINGS=examples/tensorflow/config.py flask run
```

Let's review this now!  And I'd say, let's not worry about documentation or the codecov report for now - we can fix that in separate PR's that we'll merge into this branch.

File: tests/test_picasso.py [hunk 1]
```diff
@@ -96,18 +96,21 @@ def test_saved_model(self):
         assert km.tf_predict_var is not None
 
 
-class TestTensorflowBackend:
+from picasso.ml_frameworks.tensorflow.model import TFModel
+class TFTestModel(TFModel):
+    TF_PREDICT_VAR = 'Softmax:0'
```
Comment (rhsimplex) [inline tests/test_picasso.py]: this overrides the configuration file

File: picasso/examples/tensorflow/model.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+from picasso.ml_frameworks.tensorflow.model import TFModel
+
+
+MNIST_DIM = (28, 28)
+
+
+class TensorflowMNISTModel(TFModel):
+
+    TF_INPUT_VAR = 'convolution2d_input_1:0'
```
Comment (rhsimplex) [inline picasso/examples/tensorflow/model.py]: this is configuration and shouldn't be overridden here

**Comment by jwayne** [comment]
@rhsimplex I saw you made a few updates:

1. In KerasModel, convert `self.tf_predict_var`, `self.tf_input_var` to lower case
1. In TFModel, convert class attributes `TF_PREDICT_VAR`, `TF_INPUT_VAR` to lower-case. Use 
these class attributes unless the arguments `tf_predict_var`, `tf_input_var` as passed into the constructor.

I agree with (1) because these attributes are set in an instance method (specifically, the instance constructor), which makes them class attributes. Thus, they should be lower case. Furthermore, in TFModel, these are lower-case attributes, and these attributes should be named the same way in TFModel and KerasModel. (The notion that these attributes should have the same name in TFModel and KerasModel will be clearer when we convert `tf_predict_var`, `tf_input_var` to properties in a shared interface)

I disagree with (2) for 2 reasons. Aside from breaking the convention that class attributes be upper-case, more importantly, it is bad style to override class attributes in an instance. Class attributes belong to the class, so generally users expect that they are untouched by an instance. Instances should only modify instance attributes, which they own. In this particular case, I'd recommend converting the class attributes' names back to upper case and then defining the TFModel constructor as follows:

```
def __init__(self, *args, **kwargs):
    super(TFModel, self).__init__(*args, **kwargs)
    if not hasattr(self, 'tf_predict_var'):
        self.tf_predict_var = self.TF_PREDICT_VAR
    if not hasattr(self, 'tf_input_var'):
        self.tf_input_var = self.TF_INPUT_VAR
```

Now, this is still bad style, since generally we shouldn't be using hasattr checks. But we can fix this in the separate formatting/styling PR.

**Comment by jan-xyz** [comment]
@jwayne, I'm assuming you are using PEP8 code style in the picasso.

According to PEP8 all caps is reserved for constants not for class variables. Class variables should lead with a double under score (e.g. `__tf_predict_value`) and they can be accessed by calling the base class directly since python doesn't import variables with that naming scheme automatically (e.g. `baseModel.__tf_predict_model`).

here the reference to the section in PEP8: https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables

another question: why are you not specifying the parameters for the models? In my opinion using `*args` and `**kwargs` obfuscates and reduces readability. Furthermore you could shorten the entire logic by using default parameters directly:

```python
def __init__(self, 
             tf_predict_var=BaseModel.__tf_predit_var,
             tf_input_var=BaseModel.__tf_input_var):
    super(TFModel, self).__init__(*args, **kwargs)
    self.tf_predict_var = tf_predict_var
    self.tf_input_var = tf_input_var
```

**Comment by rhsimplex** [comment]
@jwayne @jan-xyz  I agree the casing is not correct! But since I'm trying for the minimal delta for the current feature, and they are upper case and lower case throughout the code, I just changed them all to the same case to make the test pass (since the visualizations refer to the lower-case versions).

We're doing another PR with style changes on top of this one, we can fix it there if that's alright (and even merge it in to this one before merging to master).

**Comment by jwayne** [comment]
@rhsimplex Ok let's fix it in the other pr then. It was unclear to me
whether we wanted new code in this pr to follow style guides or not. To me
that made sense but if we're fixing it in the next pr anyway then I don't
care since the result is the same.

**Comment by jwayne** [comment]
@jan-xyz yes that's what I meant, thanks. But in this case TF_PREDICT_VAR,
etc are class constants. Anyway this is now a moot point since we will fix
in next pr.

**Comment by rhsimplex** [comment]
If it looks ok to you @jwayne, then I'll update the documentation and merge (or leave this branch separate for now, whichever you prefer).

**Comment by jwayne** [comment]
@rhsimplex I finished making the code style and formatting changes in the last 2 commits. I'd recommend looking at them independently. I think they should be relatively parseable, let me know and thanks!

File: picasso/__init__.py [hunk 3]
```diff
@@ -13,9 +13,23 @@
...
+                       'BACKEND_POSTPROCESSOR_PATH',
+                       'BACKEND_PROB_DECODER_NAME',
+                       'BACKEND_PROB_DECODER_PATH']
+
+if any([x in app.config.keys() for x in deprecated_settings]):
+    raise ValueError('It looks like you\'re using a deprecated'
+                     ' setting.  The settings and utility functions'
+                     ' have been changed as of version v0.2.0 (and '
+                     'you\'re using {}). Changing to the updated '
+                     ' settings is trivial: see xxxxxxx.'.format(__version__))
```
Comment (jwayne) [inline picasso/__init__.py]: remember to update this

File: picasso/__init__.py [hunk 4]
```diff
@@ -13,9 +13,23 @@
...
 
 if os.getenv('PICASSO_SETTINGS'):
     app.config.from_envvar('PICASSO_SETTINGS')
 
+deprecated_settings = ['BACKEND_PREPROCESSOR_NAME',
+                       'BACKEND_PREPROCESSOR_PATH',
+                       'BACKEND_POSTPROCESSOR_NAME',
+                       'BACKEND_POSTPROCESSOR_PATH',
+                       'BACKEND_PROB_DECODER_NAME',
+                       'BACKEND_PROB_DECODER_PATH']
```
Comment (jwayne) [inline picasso/__init__.py]: probably want to add data_dir here

Comment (rhsimplex) [inline picasso/__init__.py]: good catch

**Pull request merged**